lib/spack/docs/build_systems/cudapackage.rst

```diff
@@ -9,35 +9,120 @@
+
+  GPUs and their compute capability versions are listed at
+  https://en.wikipedia.org/wiki/CUDA#GPUs_supported.
+
+^^^^^^^^^
+Conflicts
+^^^^^^^^^
+
+Conflicts are used to prevent builds with known bugs or issues. While
+base ``CUDA`` conflicts have been included with this package, you may
+want to add more for your software.
+
+For example, if your package requires ``cuda_arch`` to be specified when
+``cuda`` is enabled, you can add the following conflict to your package
+to terminate such build attempts with a suitable message:
+
+.. code-block:: python
+
+    conflicts('cuda_arch=none', when='+cuda',
+              msg='CUDA architecture is required')
+
+Similarly, if your software does not support all versions of the property,
+you could add ``conflicts`` to your package for those versions.  For example,
+suppose your software does not work with CUDA compute capability versions
+prior to `5.0`.  You can add the following code to display a custom message
+should a user attempt such a build:
+
+.. code-block:: python
+
+    unsupported_cuda_archs = [
+        '10', '11', '12', '13',
+        '20', '21',
+        '30', '32', '35', '37'
+    ]
+    for value in unsupported_cuda_archs:
+        conflicts('cuda_arch={0}'.format(value), when='+cuda',
+                  msg='CUDA architecture {0} is not supported'.format(value))
+
+^^^^^^^
+Methods
+^^^^^^^
+
+This package provides one custom helper method, which is used to build
+standard CUDA compiler flags.
+
+**cuda_flags**
+
+    This built-in static method returns a list of command line flags 
+    for the chosen ``cuda_arch`` value(s).  The flags are intended to
+    be passed to the CUDA compiler driver (i.e., ``nvcc``).
+
+    This method must be explicitly called when you are creating the
+    arguments for your build in order to use the values.
 
 ^^^^^
 Usage
 ^^^^^
 
-In order to use it, just add another base class to your package, for example:
+This helper package can be added to your package by adding it as a base
+class of your package.  For example, you can add it to your
+:ref:`CMakePackage <cmakepackage>`-based package as follows:
 
 .. code-block:: python
+   :emphasize-lines: 1,7-16
 
-    class MyPackage(CMakePackage, CudaPackage):
+    class MyCudaPackage(CMakePackage, CudaPackage):
         ...
         def cmake_args(self):
             spec = self.spec
+            args = []
+            ...
             if '+cuda' in spec:
-                options.append('-DWITH_CUDA=ON')
-                cuda_arch = spec.variants['cuda_arch'].value
+                # Set up the cuda macros needed by the build
+                args.append('-DWITH_CUDA=ON')
+                cuda_arch_list = spec.variants['cuda_arch'].value
+                cuda_arch = cuda_arch_list[0]
                 if cuda_arch != 'none':
-                    options.append('-DCUDA_FLAGS=-arch=sm_{0}'.format(cuda_arch[0]))
+                    args.append('-DCUDA_FLAGS=-arch=sm_{0}'.format(cuda_arch))
```

Comment:
Should we be using `cuda_flags` here? The base class has a static method for this.

Reply:
The base class' `cuda_flags` returns a list of `'--generate-code arch=compute_{0},code=sm_{0} --generate-code arch=compute_{0},code=compute_{0}'.format(s)` for `s` in the list of arch's passed it.  Is that equivalent when the flags are specified this way?

Reply:
Note the `dealii` package has a comment that it has compiler errors when using the values returned from `cuda_flags` (see https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py).

Reply:
The `dihydrogen` package only uses `-arch=sm_60` (for its `CMAKE_CUDA_FLAGS` macro) when there is a single architecture with the value `auto`.